src/ir_Corona.cpp

```diff
@@ -0,0 +1,569 @@
+/// @param[in,out] data the array to be modified
+void IRCoronaAc::checksum(uint8_t* data) {
+  uint8_t pos;
+  for (uint8_t section = 0; section < kCoronaAcSections; section++) {
+    pos = section * kCoronaAcSectionBytes;
+    data[pos + kCoronaAcSectionB0Pos] = kCoronaAcSectionB0;
+    data[pos + kCoronaAcSectionB1Pos] = kCoronaAcSectionB1;
+    data[pos + kCoronaAcSectionLabelPos] = getSectionByte(section);
+    data[pos + kCoronaAcSectionD0InvPos] =
+      ~data[pos + kCoronaAcSectionD0Pos];
+    data[pos + kCoronaAcSectionD1InvPos] =
+      ~data[pos + kCoronaAcSectionD1Pos];
+  }
+}
+
+void IRCoronaAc::begin(void) { _irsend.begin(); }
+
+#if SEND_CORONA_AC
+void IRCoronaAc::send(const uint16_t repeat) {
+  _irsend.sendCoronaAc(getRaw(), kCoronaAcStateLength, repeat);
+}
+#endif  // SEND_CORONA_AC
+
+/// Get a copy of the internal state as a valid code for this protocol.
+/// @return A valid code for this protocol based on the current internal state.
+uint8_t* IRCoronaAc::getRaw(void) {
+  checksum(remote_state);  // Ensure correct settings before sending.
+  return remote_state;
+}
+
+/// Set the internal state from a valid code for this protocol.
+/// @param state A valid code for this protocol.
+void IRCoronaAc::setRaw(const uint8_t new_code[], const uint16_t length) {
+  memcpy(remote_state, new_code, std::min(length, kCoronaAcStateLength));
+}
+
+/// Set the temp in deg C.
+/// @param temp The desired temperature in Celsius.
+void IRCoronaAc::setTemp(const uint8_t temp) {
+  uint8_t degrees = std::max(temp, kCoronaAcMinTemp);
+  degrees = std::min(degrees, kCoronaAcMaxTemp);
+  setBits(&remote_state[kCoronaAcSectionD1Pos], kCoronaAcTempOffset,
+          kCoronaAcTempSize, degrees - kCoronaAcMinTemp + 1);
+}
+
+/// Get the current temperature from the internal state.
+/// @return The current temperatue in Celsius.
+uint8_t IRCoronaAc::getTemp(void) {
+  return GETBITS8(remote_state[kCoronaAcSectionD1Pos], kCoronaAcTempOffset,
+      kCoronaAcTempSize) + kCoronaAcMinTemp - 1;
+}
+
+/// Change the power setting.
+/// @param on true, the setting is on. false, the setting is off.
+/// @note If changed, setPowerToggle is also needed,
+///       unless timer is or was active
+void IRCoronaAc::setPower(const bool on) {
+  setBit(&remote_state[kCoronaAcSectionD1Pos], kCoronaAcPowerOffset, on);
+  // setting power state resets timers that would cause the state
+  if (on)
+    setOnTimer(kCoronaAcTimerOff);
+  else
+    setOffTimer(kCoronaAcTimerOff);
+}
+
+/// Get the value of the current power setting.
+/// @return true, the setting is on. false, the setting is off.
+bool IRCoronaAc::getPower(void) {
+  return GETBIT8(remote_state[kCoronaAcSectionD1Pos], kCoronaAcPowerOffset);
+}
+
+/// Change the power toggle setting.
+/// @param on true, the setting is on. false, the setting is off.
+/// @note this sets that the AC should change power,
+///       use setPower to define if the AC should end up as on or off
```

Comment:
I was confused as well, and meant to ask you about the best approach here...
If a timer is active (included in the command) then the "power bit" changes the actual state of the AC as well, however if no timer is set then the power bit alone does not change the state, and the toggle seem to be needed (at least that is the state for power on, did not test to power off without it)

don't fully understand `handleToggles` part

Reply:
One thing I have not tried is to always send it, will test that.
Not beautiful from a "emulate the remote" standpoint", but as noted before, that is not the goal either, it is to have a working control.

Reply:
Okay, I'm more enlightened, but still confused. (This is the problem with not reverse engineering the code myself ... oh what joyous problem to have ;-)
So is the "power bit" really just a "toggle the power" command/setting?

`handleToggles()` is there to try to map "brain dead" protocols that have no concept of pure "on" or "off" state in the `remote_state` to having a deterministic power state.
e.g. the `IRac` class object will know it needs to send a `toggle` when the user changes a particular setting.

i.e. Most protocols have a bit that says "the unit is on", "the unit is off". We don't need to know the previous state etc. Some, have no "on" or "off" setting, just a "change/toggle" setting. The library doesn't know if the AC is actually on or off. A user with the remote in their hand knows. So, the `IRac` class tries to keep the previous state, and determines if it needs to send a "toggle" for that setting, if the user wants to transition to a new "state".
That way we can present standard interface "user tells the class to have that setting on or off".
`handleToggle()` is the horrible hack to try to manage that logic by calculating an ephemeral state to be sent to the A/C which will change it into the new "desired" state.

Clear? (as mud?)

Reply:
> One thing I have not tried is to always send it, will test that.
> Not beautiful from a "emulate the remote" standpoint", but as noted before, that is not the goal either, it is to have a working control.

Yep. Good thing to try. I've found that "works" some times. i.e. Don't be exactly like the remote.

Correct. The end goal is to get the A/C into the same state the user wants it.

Some protocols, gulp, I've had to send two messages back-to-back to get it do all the things a user has requested. Horrible horrible, and sometimes two beeps from the A/C for the two messages, but the only way "to be sure". :-/

Reply:
Don't know if it helps, but there is this sheet: https://docs.google.com/spreadsheets/d/1zzDEUQ52y7MZ7_xCU3pdjdqbRXOwZLsbTGvKWcicqCI/ (which I forgot to set as shared Doh!)

there is 2 bits related to power
One is the "power bit" which is always consistent with if the display on the remote is on.
And then there is "power press bit" (known as the toggle in the code) which is 1 when the power button on the remote is sent.

if the remote is off, and we press timer button, it goes to "remote on" and "on timer 1h", AC is still off
Pressing timer again, the remote is of course still on, and we now have "off timer 1h" instead, and the AC has turned on, in this case the "power press bit" was never used.

however if no timer is set, and the "power press bit" is not set it does not switch to the on state even if the "power bit" is set.

I will try to hack this more to avoid the toggle and only use it internally, will probably still need `handleToggles` on swing tho

Reply:
Have tried to always send it as long as no timer is active .. first tests worked so I thought I was good and just implemented it ... as it works out (Murphy thank you!) if ac already is on then changing temp or swing or anything else have no effect....
So this means that I maybe want to send 2 commands if there is no timers...
One with the power button pressed, and one without, that should make it beep only once.

The question then is how to handle this in `getRaw()` would it be reasonable to have a flag for which packet to grab?

Reply:
Can you explain it another way? I don't think I'm getting it.
What's in each if the "sections" and when are the different sections sent?

Reply:
I get a felling there is a misunderstanding, sections are not so interesting here.
Each message has 3 sections, all are always sent (except for a special case which won't be supported other than decode, so there is always 3 sections in every sent message)
each section has 2 bytes of data
section 0 is fan, mode, temp etc
section 1 is off timer
section 2 is on timer

but maybe you mean message instead,
What I'm thinking of doing is to, as long no timer is set, send almost the same message twice.
* Once with "power button press" bit **unset**
  this should set it correctly as long as it already is the same ON/OFF state
* Then once more "power button press" bit **set**
  This should turn it ON/OFF, and do nothing if it already is correct

So will have to change state in the send function

When I say "there is no timers" I mean that they are not active, aka set to 0xFFFF

Reply:
What does the real remote do/send?
(and how did your trial go with having "power button" always set to `1`?

Reply:
> The question then is how to handle this in getRaw() would it be reasonable to have a flag for which packet to grab?

SO, I think I now understand enough to answer this. i.e. why it was being asked.
`getRaw()`s job is simple. Return the internal state of the remote such that it can be used to send a message with `sendCorona()` etc.

I _think_ there is one protocol that has two different internal states (one message has settings, the other has timer/clock functions etc). I think it defaults to what ever setting (or setRaw()) was set last for it's `getRaw()`.

If you think you need to send two messages, to set the a/c into a known state, then I'd probably put that in `IRCoronaAc::send()`'s code, and leave the result of `getRaw()` as is. Not. you don't have to use `getRaw()` to feed `sendCorona()` if you don't want to. You're free to do what ever you think you need/is best etc.

Reply:
TL;DR: I won't add parameters to `getRaw()` myself, but there is not reason you can't. Just make sure the parameter has a default, so that a simple `getRaw()` call (no args) still works as a user would expect.

Reply:
The real remote only sends "power press" when that button is pressed.
The only other way to toggle the AC to on is to do so by cycling thru timers, first on timer becomes active, once more the AC turns on (no timers active), once more off timer goes active
* Turning On (power press)
```
Mesg Desc.: Power: On, Power Button: On, Mode: 3 (Fan), Temp: 20C, Fan: 1 (Low), Swing(V) Toggle: Off, Econo: Off, On Timer: Off, Off Timer: Off
uint8_t state[21] = {
0x28, 0x61, 0x3D, 0x11, 0xEE, 0xF4, 0x0B,
0x28, 0x61, 0x6D, 0xFF, 0x00, 0xFF, 0x00,   // No On Timer
0x28, 0x61, 0xCD, 0xFF, 0x00, 0xFF, 0x00};  // No Off Timer
```

* Turning Off (power press)
```
Mesg Desc.: Power: Off, Power Button: On, Mode: 3 (Fan), Temp: 20C, Fan: 1 (Low), Swing(V) Toggle: Off, Econo: Off, On Timer: Off, Off Timer: Off
uint8_t state[21] = {
0x28, 0x61, 0x3D, 0x11, 0xEE, 0xE4, 0x1B,
0x28, 0x61, 0x6D, 0xFF, 0x00, 0xFF, 0x00,   // No On Timer
0x28, 0x61, 0xCD, 0xFF, 0x00, 0xFF, 0x00};  // No Off Timer
```

* Off pressing timer
```
Mesg Desc.: Power: On, Power Button: Off, Mode: 3 (Fan), Temp: 20C, Fan: 1 (Low), Swing(V) Toggle: Off, Econo: Off, On Timer: 06:00, Off Timer: Off
uint8_t state[21] = {
0x28, 0x61, 0x3D, 0x11, 0xEE, 0xD4, 0x2B,
0x28, 0x61, 0x6D, 0x30, 0xCF, 0x2A, 0xD5,   // 6 hours
0x28, 0x61, 0xCD, 0xFF, 0x00, 0xFF, 0x00};  // No Off Timer
```

* pressing timer, ac turns on
```
Mesg Desc.: Power: On, Power Button: Off, Mode: 3 (Fan), Temp: 20C, Fan: 1 (Low), Swing(V) Toggle: Off, Econo: Off, On Timer: Off, Off Timer: Off
uint8_t state[21] = {
0x28, 0x61, 0x3D, 0x11, 0xEE, 0xD4, 0x2B,
0x28, 0x61, 0x6D, 0xFF, 0x00, 0xFF, 0x00,   // No On Timer
0x28, 0x61, 0xCD, 0xFF, 0x00, 0xFF, 0x00};  // No Off Timer
```

* pressing timer again (off timer becomes active)
```
Mesg Desc.: Power: On, Power Button: Off, Mode: 3 (Fan), Temp: 20C, Fan: 1 (Low), Swing(V) Toggle: Off, Econo: Off, On Timer: Off, Off Timer: 01:00
uint8_t state[21] = {
0x28, 0x61, 0x3D, 0x11, 0xEE, 0xD4, 0x2B,
0x28, 0x61, 0x6D, 0xFF, 0x00, 0xFF, 0x00,   // No On Timer
0x28, 0x61, 0xCD, 0x08, 0xF7, 0x07, 0xF8};  // 1 hour
```

If the AC is Off I can send it with power button set, and it turns on.
But sending any command when it already is On then there is **no response** from the AC (it usually beeps)

Therefore I think that, sending once with power button **unset** would change any state it has **if it is on**.
And then send once more with power button **set** should turn it on if it already isn't

Reply:
> The real remote only sends "power press" when that button is pressed.
> The only other way to toggle the AC to on is to do so by cycling thru timers, first on timer becomes active, once more the AC turns on (no timers active), once more off timer goes active

Sigh ... sometimes I'd like met some of these IR remote engineers in a dark ally .. and ask them what the hell were they thinking?!

Reply:
haha

I'm testing this now: (and some related cleanup)
```
void IRCoronaAc::send(const uint16_t repeat) {
  // if no timer, always send once without power press
  if (!getOnTimer() || !getOffTimer()) {
    setPowerButton(false);
    _irsend.sendCoronaAc(getRaw(), kCoronaAcStateLength, repeat);
    // and then with power press
    setPowerButton(true);
  }
  _irsend.sendCoronaAc(getRaw(), kCoronaAcStateLength, repeat);
}
```

Will try to cycle thru enough testcases to be sure that this works.

Reply:
> If the AC is Off I can send it with power button set, and it turns on.
> But sending any command when it already is On then there is **no response** from the AC (it usually beeps)

Um .. so if Power = On && Power Button = Off, it won't accept the command/setting changes?

Reply:
> > If the AC is Off I can send it with power button set, and it turns on.
> > But sending any command when it already is On then there is **no response** from the AC (it usually beeps)
> 
> Um .. so if Power = On && Power Button = Off, it won't accept the command/setting changes?

If you are talking about bits in the message, then if the AC unit already is on than that will work.
If the AC is in off mode it won't accept it, like an xor
AC On, sending a command with On bit set, and Power Press bit results in nothing
AC On, sending a command with On bit set, without Power Press Bit is ok
AC Off, sending a command with On bit set, without Power Press Bit results in nothing
AC Off, sending a command with On bit set, and Power Press Bit is ok (turns it on)
Will extend the sheet with a full list of tests for this, just to be sure about all states

Reply:
> If the AC is in off mode it won't accept it, like an xor
AC On, sending a command with On bit set, and Power Press bit results in nothing
AC On, sending a command with On bit set, without Power Press Bit is ok
AC Off, sending a command with On bit set, without Power Press Bit results in nothing
AC Off, sending a command with On bit set, and Power Press Bit is ok (turns it on)

I think that deserves to be a comment in the code.

(I really hate A/C vendor logic some times, especially in the last 48 hrs)

Reply:
You are not alone...
So full specification of all the different bits (related to bower) and the result:
https://docs.google.com/spreadsheets/d/1zzDEUQ52y7MZ7_xCU3pdjdqbRXOwZLsbTGvKWcicqCI/edit#gid=496223625
It mostly confirms what I was thinking before, unless there is a timer active, I need to send things twice, since the AC active state is unknown. - and that seems to work, will test a bit more tho.

Reply:
gah. Sounds horrible. You've got the unit, so I'll leave the flavour of hacks up to you.